File: ios/App/App/HabitWidgetPlugin.swift
import Capacitor
import WidgetKit

@objc(HabitWidgetPlugin)
public class HabitWidgetPlugin: CAPPlugin {

    private let suiteName = "group.com.trackr.app"

    @objc func syncHabits(_ call: CAPPluginCall) {
        guard let habitsJson = call.getString("habitsJson") else {
            call.reject("habitsJson required")
            return
        }

        let userId       = call.getString("userId")       ?? ""
        let idToken      = call.getString("idToken")      ?? ""
        let refreshToken = call.getString("refreshToken") ?? ""
        let apiKey       = call.getString("apiKey")       ?? ""
        let projectId    = call.getString("projectId")    ?? ""
        let tokenExpiry  = call.getDouble("tokenExpiry")  ?? 0.0

        guard let defaults = UserDefaults(suiteName: suiteName) else {
            call.reject("App Group not configured — add 'group.com.trackr.app' in Xcode Signing & Capabilities")
            return
        }

        defaults.set(habitsJson,   forKey: "habitsJson")
        defaults.set(userId,       forKey: "userId")
        defaults.set(idToken,      forKey: "idToken")
        defaults.set(refreshToken, forKey: "refreshToken")
        defaults.set(apiKey,       forKey: "apiKey")
        defaults.set(projectId,    forKey: "projectId")
        defaults.set(tokenExpiry,  forKey: "tokenExpiry")
        defaults.synchronize()

        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }

        call.resolve()
    }

    @objc func triggerWidgetUpdate(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve()
    }
}

File: android/app/src/main/java/com/trackr/app/HabitWidgetPlugin.kt
package com.trackr.app

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

const val WIDGET_PREFS_NAME = "com.trackr.app.widget_prefs"

@CapacitorPlugin(name = "HabitWidgetPlugin")
class HabitWidgetPlugin : Plugin() {

    @PluginMethod
    fun syncHabits(call: PluginCall) {
        val habitsJson   = call.getString("habitsJson")   ?: run { call.reject("habitsJson required"); return }
        val userId       = call.getString("userId")       ?: ""
        val idToken      = call.getString("idToken")      ?: ""
        val refreshToken = call.getString("refreshToken") ?: ""
        val apiKey       = call.getString("apiKey")       ?: ""
        val projectId    = call.getString("projectId")    ?: ""
        val tokenExpiry  = call.getLong("tokenExpiry", 0L) ?: 0L

        val prefs = context.getSharedPreferences(WIDGET_PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().apply {
            putString("habitsJson",   habitsJson)
            putString("userId",       userId)
            putString("idToken",      idToken)
            putString("refreshToken", refreshToken)
            putString("apiKey",       apiKey)
            putString("projectId",    projectId)
            putLong("tokenExpiry",    tokenExpiry)
            apply()
        }

        refreshWidgets()
        call.resolve()
    }

    @PluginMethod
    fun triggerWidgetUpdate(call: PluginCall) {
        refreshWidgets()
        call.resolve()
    }

    private fun refreshWidgets() {
        val manager = AppWidgetManager.getInstance(context)
        val ids = manager.getAppWidgetIds(ComponentName(context, HabitWidgetReceiver::class.java))
        if (ids.isNotEmpty()) {
            manager.notifyAppWidgetViewDataChanged(ids, android.R.id.list)
        }
    }
}
